### PR TITLE
ext_authz: option for clearing route cache of authorized requests

### PR DESCRIPTION
**Pull request by gsagula**
Description:
This PR adds a configurable option to `ext_authz` that allows the filter clearing the route cache. This is particularly useful when the authorization response headers are intended to influence the route decision of an authorized request.

Risk Level: Low
Testing: Yes
Docs Changes: Yes
Release Notes: Yes
Fixes: #6481

### CODE REVIEW

File: api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto [hunk 1]
```diff
@@ -53,6 +53,17 @@ message ExtAuthz {
 
   // Enables filter to buffer the client request body and send it within the authorization request.
   BufferSettings with_request_body = 5;
+
+  // Clears route cache in order to allow the external authorization service to correctly affect
+  // routing decisions. Filter clears all cached routes when:
+  //
+  // 1. The field is set to *true*.
+  //
```
Comment (dio) [inline api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto]: Superfluous newlines?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc [hunk 2]
```diff
@@ -500,6 +500,225 @@ TEST_F(HttpFilterTest, HeaderOnlyRequestWithStream) {
...
+  )EOF");
+
+  prepareCheck();
+
+  EXPECT_CALL(*client_, check(_, _, testing::A<Tracing::Span&>()))
+      .WillOnce(
+          WithArgs<0>(Invoke([&](Filters::Common::ExtAuthz::RequestCallbacks& callbacks) -> void {
+            request_callbacks_ = &callbacks;
+          })));
+  EXPECT_CALL(filter_callbacks_, clearRouteCache());
```
Comment (dio) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc]: To make it clear, let's add `.Times(1)`?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc [hunk 3]
```diff
@@ -500,6 +500,225 @@ TEST_F(HttpFilterTest, HeaderOnlyRequestWithStream) {
...
+
+  prepareCheck();
+
+  Filters::Common::ExtAuthz::Response response{};
+  response.status = Filters::Common::ExtAuthz::CheckStatus::Denied;
+  response.status_code = Http::Code::Unauthorized;
+  response.headers_to_add = Http::HeaderVector{{Http::LowerCaseString{"foo"}, "bar"}};
+  auto response_ptr = std::make_unique<Filters::Common::ExtAuthz::Response>(response);
+
+  EXPECT_CALL(*client_, check(_, _, _))
```
Comment (dio) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc]: Sorry, a question: why we don't expect to call `check(_, _, testing::A<Tracing::Span&>())` here?

File: source/extensions/filters/http/ext_authz/ext_authz.cc [hunk 4]
```diff
@@ -182,6 +182,12 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
     // Only send headers if the response is ok.
     if (response->status == CheckStatus::OK) {
       ENVOY_STREAM_LOG(trace, "ext_authz filter added header(s) to the request:", *callbacks_);
+      if (config_->clearRouteCache() &&
+          (!response->headers_to_add.empty() || !response->headers_to_append.empty())) {
```
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc]: Sorry, for my own education, what is the relation between the existence of `headers_to_add` or `headers_to_append` with the decision of clearing route cache?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc (hunk 3 above)
Comment (gsagula) [reply test/extensions/filters/http/ext_authz/ext_authz_test.cc]: We probably do. I just copied it from some other place :)

File: source/extensions/filters/http/ext_authz/ext_authz.cc (hunk 4 above)
Comment (gsagula) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Sure, being able to route a request based on authorization response metadata is the whole point of clearing the cache. If some particular request has no metadata to be added/appended, just leave the route cache how it is. At least that is how I understand it.

File: api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto [hunk 5]
```diff
@@ -53,6 +53,15 @@ message ExtAuthz {
 
   // Enables filter to buffer the client request body and send it within the authorization request.
   BufferSettings with_request_body = 5;
+
+  // Clears route cache in order to allow the external authorization service to correctly affect
+  // routing decisions. Filter clears all cached routes when:
+  //
+  // 1. The field is set to *true*.
+  // 2. The status returned from the authorization service is a HTTP 200 or gRPC 0.
+  // 3. One or more *authorization response headers* are addressed to the upstream.
```
Comment (mattklein123) [inline api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto]: Sorry what does this mean? Can you clarify?

File: docs/root/intro/version_history.rst [hunk 6]
```diff
@@ -3,6 +3,7 @@ Version history
 
 1.11.0 (Pending)
 ================
+* ext_authz: added option to `ext_authz` that allows the filter clearing route cache.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: alpha order

File: api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto (hunk 5 above)
Comment (gsagula) [reply api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto]: I rephrased it.

**Pull request merged**